Code/GraphMol/SynthonSpaceSearch/SynthonSpaceSearcher.cpp

```diff
@@ -0,0 +1,265 @@
+}
+
+std::unique_ptr<ROMol> SynthonSpaceSearcher::buildAndVerifyHit(
+    const std::unique_ptr<SynthonSet> &reaction,
+    const std::vector<size_t> &synthNums,
+    std::set<std::string> &resultsNames) const {
+  const auto prodName = reaction->buildProductName(synthNums);
+
+  std::unique_ptr<ROMol> prod;
+  if (resultsNames.insert(prodName).second) {
+    if (resultsNames.size() < static_cast<size_t>(d_params.hitStart)) {
+      return prod;
+    }
+    prod = reaction->buildProduct(synthNums);
+
+    // Do a final check of the whole thing.  It can happen that the
+    // fragments match synthons but the final product doesn't match.
+    // A key example is when the 2 synthons come together to form an
+    // aromatic ring.  For substructure searching, an aliphatic query
+    // can match the aliphatic synthon so they are selected as a hit,
+    // but the final aromatic ring isn't a match.
+    // E.g. Cc1cccc(C(=O)N[1*])c1N=[2*] and c1ccoc1C(=[2*])[1*]
+    // making Cc1cccc2c(=O)[nH]c(-c3ccco3)nc12.  The query c1ccc(CN)o1
+    // when split is a match to the synthons (c1ccc(C[1*])o1 and [1*]N)
+    // but the product the hydroxyquinazoline is aromatic, at least in
+    // the RDKit model so the N in the query doesn't match.
+    if (!verifyHit(*prod)) {
+      prod.reset();
+    }
+  }
+  if (prod) {
+    prod->setProp<std::string>(common_properties::_Name, prodName);
+  }
+  return prod;
+}
+
+void SynthonSpaceSearcher::buildHits(
+    std::vector<SynthonSpaceHitSet> &hitsets, const size_t totHits,
+    std::vector<std::unique_ptr<ROMol>> &results) const {
+  if (hitsets.empty()) {
+    return;
+  }
+  std::sort(
+      hitsets.begin(), hitsets.end(),
+      [](const SynthonSpaceHitSet &hs1, const SynthonSpaceHitSet &hs2) -> bool {
+        if (hs1.reactionId == hs2.reactionId) {
+          return hs1.numHits < hs2.numHits;
+        }
+        return hs1.reactionId < hs2.reactionId;
+      });
+  // Keep track of the result names so we can weed out duplicates by
+  // reaction and synthons.  Different splits may give rise to the same
+  // synthon combination.  This will keep the same molecule produced via
+  // different reactions which I think makes sense.  The resultsNames will
+  // be accumulated even if the molecule itself doesn't make it into the
+  // results set, for example if it isn't a random selection or it's
+  // outside maxHits or hitStart.
+  std::set<std::string> resultsNames;
+
+  if (d_params.randomSample) {
+    buildRandomHits(hitsets, totHits, resultsNames, results);
+  } else {
+    buildAllHits(hitsets, resultsNames, results);
+  }
+}
+
+namespace {
+void sortHits(std::vector<std::unique_ptr<ROMol>> &hits) {
+  if (!hits.empty() && hits.front()->hasProp("Similarity")) {
+    std::sort(hits.begin(), hits.end(),
+              [](const std::unique_ptr<ROMol> &lhs,
+                 const std::unique_ptr<ROMol> &rhs) {
+                const auto lsim = lhs->getProp<double>("Similarity");
```

Comment:
for the future: if profiling shows that we're spending significant time in this function, we should explore an alternative to storing the similarities as propertiesl they are convenient, but not super fast

Reply:
I thought about deriving a SimilaritySearchResults class but decided to take the path of least resistance for now.